Alert user upon switching to unconnected account

Gudahtt opened the pull request:
An alert is now shown when the user switches from an account that is connected to the active tab to an account that is not connected. The alert prompts the user to dismiss the alert or connect the account they're switching to.

The "loading" state is handled by disabling the buttons, and the error state is handled by displaying a generic error message and disabling the connect button.

The new reducer for this alert has been created with `createSlice` from the Redux Toolkit. This utility is recommended by the Redux team, and represents a new style of writing reducers that I hope we will use more in the future (or at least something similar). `createSlice` constructs a reducer, actions, and action creators automatically. The reducer is constructed using their `createReducer` helper, which uses Immer to allow directly mutating the state in the reducer but exposing these changes as immutable.

Gudahtt committed:
Alert user upon switching to unconnected account

An alert is now shown when the user switches from an account that is
connected to the active tab to an account that is not connected. The
alert prompts the user to dismiss the alert or connect the account
they're switching to.

The "loading" state is handled by disabling the buttons, and the error
state is handled by displaying a generic error message and disabling
the connect button.

The new reducer for this alert has been created with `createSlice` from
the Redux Toolkit. This utility is recommended by the Redux team, and
represents a new style of writing reducers that I hope we will use more
in the future (or at least something similar). `createSlice` constructs
a reducer, actions, and action creators automatically. The reducer is
constructed using their `createReducer` helper, which uses Immer to
allow directly mutating the state in the reducer but exposing these
changes as immutable.

Gudahtt commented:
I'll wait on #8345 before fixing the broken integration tests, as I'll have one less state file to modify after that's merged.

Gudahtt commented:
<details>
<summary>Screenshots:</summary>

Here's the alert:

![normal](https://user-images.githubusercontent.com/2459287/79519866-8023cb80-802b-11ea-9adf-5ef96c1f74d8.png)

Here is the error state:

![error](https://user-images.githubusercontent.com/2459287/79519869-81ed8f00-802b-11ea-9fc3-c25bd6711bad.png)

</details>

Gudahtt commented on ui/app/ducks/alerts/unconnected-account.js:
```diff
@@ -0,0 +1,101 @@
+  },
+  extraReducers: {
+    [actionConstants.UPDATE_METAMASK_STATE]: (state, action) => {
+      // lose the alert if the account is switched while it's open
```
Nit: typo, this should say `close`, not `lose`
```suggestion
      // close the alert if the account is switched while it's open
```

Gudahtt commented on ui/app/components/app/alerts/alerts.scss:
```diff
@@ -0,0 +1 @@
+@import './unconnected-account-alert/unconnected-account-alert.scss'
```
Nit: missing semi-colon at the end of this line

```suggestion
@import './unconnected-account-alert/unconnected-account-alert.scss';
```

Gudahtt commented on ui/app/components/app/alerts/unconnected-account-alert/unconnected-account-alert.js:
```diff
@@ -0,0 +1,64 @@
+        </>
+      )}
+      footerClassName="unconnected-account-alert__footer"
+    >
+    </Popover>
```
Nit: This should be a self-closing, since there is nothing being passed in as children.
```suggestion
    />
```

Gudahtt commented on ui/app/components/app/alerts/alerts.js:
```diff
@@ -0,0 +1,19 @@
+import { alertIsOpen as unconnectedAccountAlertIsOpen } from '../../../ducks/alerts/unconnected-account'
+
+const Alerts = () => {
+  const isOpen = useSelector(unconnectedAccountAlertIsOpen)
```
Nit: this is a poor choice in variable name, since I'll be adding similar variables for each other alert as well (they can't all be called `isOpen`).

Gudahtt commented on ui/app/ducks/alerts/unconnected-account.js:
```diff
@@ -0,0 +1,101 @@
+      state.address = action.payload
+    },
+  },
+  extraReducers: {
```
This doesn't need to block review, but I'm working on an improvement to this reducer that would remove the necessity to cache the current selected address. This approach of caching the relevant background state and watching for state updates doesn't scale well when you want to watch a larger set of things, and keeping redundant state in reducers is generally not advised. This was particularly noticeable while I was working on the next alert, which also needs to watch for changes to the current active tab.

Gudahtt replied:
This has been completed in bf55bbaee0f75d0230a16525cc6073a887dbc65b

rekmarks reviewed (changes requested):
Very clean and crisp! A pleasure to review. Excited to see Immer being used, even if indirectly.

One in-line comment, and also some feedback about styling/UX:

![image](https://user-images.githubusercontent.com/25517051/80432751-fd600380-88a9-11ea-9565-14bf9a204af1.png)

<details>
<summary><strong>Full Screenshot for Scale</strong></summary>
<img src="https://user-images.githubusercontent.com/25517051/80433263-4cf2ff00-88ab-11ea-956b-1be28b0cc43e.png"/>
</details>

- The buttons appear disproportionately large to me, relative to the size of the popover. Could we make them smaller?
  - Related: Where are the designs?
- In testing this, I quickly got very annoyed with this popup. Can we add a "Don't show this again" checkbox?

Gudahtt commented on test/unit/ui/app/actions.spec.js:
```diff
@@ -331,7 +331,7 @@ describe('Actions', function () {
 
     it('calls importAccountWithStrategies in background', async function () {
-      const store = mockStore()
+      const store = mockStore({ metamask: devState })
```
This was added so that these [two selectors would ](https://github.com/MetaMask/metamask-extension/pull/8312/files#diff-eebb93d0666faf7aa39f85d9f7efa804R1204) not blow up. I should replace this with a more minimal state.

rekmarks replied:
Got it. Definitely won't block on that, whatever you do!

Gudahtt replied:
I've replaced these with a more minimal state (just the selected address).

Gudahtt commented:
Yeah, the buttons do look a bit large. I think those sizes are being dictated by our default button styles? I'll take a look. We have another card for updating the button styles across the extension to be rounded, as shown in the Figma designs, but that wasn't going to be in v8. Maybe I can tweak them to be a bit smaller though.

The designs are [here](https://www.figma.com/file/CvD7nhQVeSEB1QuSy3XhtY/oCap-%2F-Login-Per-Site?node-id=1929%3A139).

The "Don't show this again" checkbox was going to be in a follow-up PR; there's a separate card for that on Trello I believe.

rekmarks commented:
@Gudahtt word, sounds good! Just me when you've decided what to do with the button styles.

Oh one more thing: doing any e2e tests for this in a follow-up, or what's the plan for that?

rekmarks reviewed (dismissed):
LGTM!

Buttons and checkbox can be fixed in one or more follow-up PRs.

Gudahtt committed:
Use `SELECTED_ADDRESS_CHANGED` state instead of cached address

The `unconnected-account` reducer no longer caches the selected address
when the alert was opened. Instead a new event was added for when the
selected address changes (`SELECTED_ADDRESS_CHANGED`). Now that we know
exactly when the selected address changes, we don't need to cache the
original account for comparison anymore.

Gudahtt committed:
Update button styles to reduce size

The buttons are now closer in size to the buttons in the designs. The
button styles will be updated to match the designs more closely in a
separate PR.

Gudahtt commented:
Two updates have been made:

* Use `SELECTED_ADDRESS_CHANGED` state instead of cached address

  The `unconnected-account` reducer no longer caches the selected address when the alert was opened. Instead a new event was added for when the selected address changes (`SELECTED_ADDRESS_CHANGED`). Now that we know exactly when the selected address changes, we don't need to cache the original account for comparison anymore.

* Update button styles to reduce size

  The buttons are now closer in size to the buttons in the designs. The button styles will be updated to match the designs more closely in a separate PR.

<details>
<summary>Screenshots of updated buttons:</summary>

![updated-buttons](https://user-images.githubusercontent.com/2459287/80543347-0a9df080-8985-11ea-825c-ac544e1a593c.png)

</details>

Gudahtt commented:
@rekmarks I started trying to write an e2e test for this, but it was a bit challenging because we can't actually test the real popup with Selenium. It doesn't allow interacting with browser-things outside of the page (even the keyboard shortcut is intercepted). Even if it did allow it, there's no way to focus on the popup either.

In other tests we fake the popup UI by navigating to `popup.html` and making the window vagely popup-sized. The `activeTab` state still won't be set in that case though, and I rely upon that for this alert. We'll need to find a way to fake the `activeTab` state - then this should be testable.

rekmarks reviewed (dismissed):
LGTM!

Nice fix for the buttons and selected address state.

rekmarks reviewed (approved):
Re-approve after rebase.

whymarrh reviewed (approved):
This looks good to me! I didn't test it out locally but it looks like @rekmarks did.
